qiskit/pulse/ir/ir.py

```diff
@@ -20,279 +20,107 @@
+        for elm in self.elements():
+            if isinstance(elm, IrBlock):
+                inst_targets |= elm.inst_targets
+            else:
+                inst_targets.add(elm.inst_target)
+        return inst_targets
 
     @property
-    def duration(self) -> int:
-        """The duration of the instruction (in terms of system ``dt``)."""
-        return self._instruction.duration
+    def sequence(self) -> PyDAG:
+        """Return the DAG sequence of the IrBlock"""
+        return self._sequence
 
-    @initial_time.setter
-    def initial_time(self, value: int):
-        """Set ``initial_time``
+    def append(self, element: IrBlock | Instruction) -> None:
+        """Append element to the IrBlock"""
+        new_node_id = self._sequence.add_node(element)
+        if isinstance(element, IrBlock):
+            self._children.append(new_node_id)
 
-        Args:
-            value: The integer value of ``initial_time``.
-
-        Raises:
-            PulseError: if ``value`` is not ``None`` and not non-negative integer.
-        """
-        if not isinstance(value, (int, np.integer)) or value < 0:
-            raise PulseError("initial_time must be a non-negative integer")
-        self._initial_time = value
+    def elements(self) -> list[IrBlock | Instruction]:
+        """Return a list of all elements in the IrBlock"""
+        return self._sequence.nodes()[2:]
 
-    def shift_initial_time(self, value: int) -> None:
-        """Shift ``initial_time``
+    def scheduled_elements(self) -> list[list[int | None, IrBlock | Instruction]]:
+        """Return a list of scheduled elements.
 
-        Shifts ``initial_time`` to ``initial_time+value``.
-
-        Args:
-            value: The integer value by which ``initial_time`` is to be shifted.
-
-        Raises:
-            PulseError: If the instruction is not scheduled.
+        Each element in the list is [initial_time, element].
         """
-        if self.initial_time is None:
-            raise PulseError("Can not shift initial_time of an untimed element")
-
-        # validation of new initial_time is done in initial_time setter.
-        self.initial_time = self.initial_time + value
-
-    def __eq__(self, other: "IrInstruction") -> bool:
-        """Return True iff self and other are equal
-
-        Args:
-            other: The IR instruction to compare to this one.
-
-        Returns:
-            True iff equal.
-        """
-        return (
-            type(self) is type(other)
-            and self._instruction == other.instruction
-            and self._initial_time == other.initial_time
-        )
-
-    def __repr__(self) -> str:
-        """IrInstruction representation"""
-        return f"IrInstruction({self.instruction}, t0={self.initial_time})"
-
+        return [
+            [self._time_table.get(ni, None), self._sequence.get_node_data(ni)]
+            for ni in self._sequence.node_indices()
+            if ni not in (0, 1)
+        ]
```

Comment:
```suggestion
        def _to_time_inst_tuple(n_index):
            instruction = self._sequence.get_node_data(ni)
            if t0 := self._time_table.get(n_index, None) is not None:
                return t0 + self.time_offset, instruction
            return None, instruction
        
        return list(
            map(
                _to_time_inst_tuple, 
                filter(lambda i: i not in (0, 1), self._sequence.node_indices()),
            )
        )
```
You need to consider the time offset. I prefer tuple form for time instruction duo.

Reply:
I opted not to do that, because it causes duplication where the initial time of sub blocks is stored in two places - the time table of the parent block, and the sub block itself. In general, I think it's better to have the initial time directly in the nodes. Eventually, this is where you want this the data to live - what is the instruction, and what time it should be applied.

(In the current implementation I took care of the offset in the `flatten` method. I left the sub-blocks unaware of their relative timing until that point).

Reply:
> the initial time of sub blocks is stored in two places - the time table of the parent block, and the sub block itself. 

Hmm it depends how we calculate initial time. I assume an outer block just calls sub block's `schedule_elements` and extend, but if you assume sub block's element time `t0 = outer block t0 + sub block element t0`, then current logic looks reasonable. 

>(In the current implementation I took care of the offset in the flatten method. I left the sub-blocks unaware of their relative timing until that point). 

I prefer this `schedule_elements` returns elements in _flattened_ fashion, since lower level representation may not have notion of context and I don't think of any use case requiring time information but context must be preserved. You can also add `flatten` argument to this method (I prefer it defaults to True).

> I think it's better to have the initial time directly in the nodes

I'm against this design, based on the following rule:

_Same command (i.e. Instruction tied to channel) should be equivalent object._

When you attach time information directly to the command, for example, two play instructions issued at different time must become different objects. To avoid this, we need a wrapper object, e.g. scheduled command, and adopt composite pattern as you first implemented. However, this results in complex class dependency and class instantiation overhead as the number of instruction increases.